idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt

```diff
@@ -0,0 +1,132 @@
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.constants.ConstantValue
+import org.jetbrains.kotlin.resolve.constants.evaluate.ConstantExpressionEvaluator
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameUnsafe
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+
+class ReplaceManualRangeWithIndicesCallsInspection : AbstractKotlinInspection() {
+
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+
+        return expressionVisitor { expression ->
+            if (expression !is KtBinaryExpression && expression !is KtDotQualifiedExpression) return@expressionVisitor
+
+            val fqName = expression.getCallableDescriptor()?.fqNameUnsafe?.asString() ?: return@expressionVisitor
+
+            if (!fqName.matches(REGEX_RANGE_TO) && !fqName.matches(REGEX_UNTIL)) return@expressionVisitor
+
+            when {
+                fqName.matches(REGEX_RANGE_TO) -> visitRangeToExpression(expression, holder)
+                fqName.matches(REGEX_UNTIL) -> visitUntilExpression(expression, holder)
+                else -> return@expressionVisitor
+            }
+
+        }
+    }
+
+    private fun visitRangeToExpression(expression: KtExpression, holder: ProblemsHolder) {
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeMinusOne() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
+        if (rightValue != 0) return
+
+
+        holder.registerProblem(
+            expression,
+            "'rangeTo or the '..' call should be replaced with .indices call",
+            ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
+            ReplaceManualRangeWithIndicesCallQuickFix()
+        )
+
+    }
+
+    private fun visitUntilExpression(expression: KtExpression, holder: ProblemsHolder) {
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeOrLength() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
+        if (rightValue != 0) return
+
+
+        holder.registerProblem(
+            expression,
+            "'rangeTo or the '..' call should be replaced with .indices call",
+            ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
+            ReplaceManualRangeWithIndicesCallQuickFix()
+        )
+
+    }
+
+
+    private fun KtExpression.isSizeMinusOne(): Boolean {
+        if (this !is KtBinaryExpression) return false
+        if (operationToken != KtTokens.MINUS) return false
+
+        val leftValue = left?.isSizeOrLength() ?: return false
+        val constantValue = right?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return false
```

Comment:
Do you need `toInt` here? I think you can use `rightValue == 1`.

Reply:
Not sure, but took that from another inspection "ReplaceRangetoWithUntilInspection" and it needs to return the same thing

Reply:
OK, from ReplaceRangetoWithUntilInspection.